core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -183,7 +183,8 @@ abstract class AbstractFetcherThread(name: String,
                     processPartitionData(topicPartition, currentPartitionFetchState.fetchOffset, partitionData)
 
                     val validBytes = records.validBytes
-                    if (validBytes > 0) {
+                    // ReplicaDirAlterThread may have removed topicPartition from the partitionStates after processing the partition data
+                    if (validBytes > 0 && partitionStates.contains(topicPartition)) {
```

Comment:
I am not sure I understand the problem here. It seems that current error logging in AbstractFetcherThread is technically correct. For example, the `Error to broker` in line 220 is correct because `sourceBroker.id` still refers to the broker from which the data is fetched, for both ReplicaFetcherThread and ReplicaAlterDirThread. Furthermore, the log4j error logging already identifies the class (either ReplicaFetcherThread or ReplicaAlterDirThread) for those log statements in the AbstractFetcherThread, thus there is probably no need to identify in the logging message whether it is for ReplicaFetcherThread or ReplicaAlterDirThread.

Does this make sense?

Reply:
My point is that for ReplicaAlterDirThread, the logging "to broker" doesn't convey much info since it always moves data within the same broker. When there is an error, it's more useful to know the source log dir from which the future replica is copying.

Reply:
@junrao I see. Currently AbstractFetcherThread can not get the log directory for partition in general because ConsumerFetcherThread, which also extends AbstractFetcherThread, does not have log directory for partition.

I think the existing implementation probably already provides log directory of the partition in the error log when this information is needed. The idea is that we need to know the log directory of the partition in the error log only if this is a log directory failure. But if this is the case, the original exception should have specified the log directory of the partition.

Does this sound reasonable?